### PR TITLE
Task/Create RZNevada CI

### PR DESCRIPTION
**Pull request by kab163**
# Creating rznevada Gitlab CI pipeline

- This PR is a new CI feature to test HIP builds on rznevada with cce and clang (so far)
- It pulls in RADIUSS changes recently merged for spack configs specific to toss4 and then uses that to create specs for HIP builds on rznevada

How to trigger a RZNevada CI pipeline:
1. On your local branch, make any changes needed and push to the normal RAJAPerf Gitlab repo
2. Push to the new RAJAPerf RZ Gitlab repo with `git push rz-gitlab my_branch_name`
3. Check the status of the pipeline, by looking at it here: https://rzlc.llnl.gov/gitlab/raja/rajaperf/-/pipelines

### CODE REVIEW

**Comment by kab163** [comment]
I need to see how to set up the CI to run on rznevada, but otherwise these jobs work (manual builds locally).

**Comment by kab163** [comment]
Only things left to do here are:
- [ ] Figure out Gitlab mirroring for RZ
- [ ] If a gcc build is needed, ask AMD folks how to fix unknown flag options
- [ ] Make sure clang compiler version is set up ok

File: .gitlab/corona-templates.yml
Comment (adrienbernede) [inline .gitlab/corona-templates.yml:22]: This rule applies to all corona jobs, so I suggest the following strategy for rules:
- first: define all rules that prevent those jobs from running
- then: define all rules specific to one job (here release_resources)
- finally: conclude with a default rules that allows the job to run `on_success` in all other circumstances.

This is because gitlab won’t run the job if there is not true rule and because it is easier to stack "when: never" rules.

File: .gitlab/corona-templates.yml
Comment (adrienbernede) [inline .gitlab/corona-templates.yml:23]: here is my suggestion:
```
  rules:
    - if: '$ON_CORONA == "OFF"'
      when: never   # no jobs on corona if ON_CORONA == OFF
    - if: '$LC_ZONE ~= "CZ"' #run only on CZ ...
      when: never   # no jobs on corona if not on CZ
    - if: '$CI_JOB_NAME =~ /release_resources/'
      when: always   # release_resources jobs should run whatever the previous stage status
    - when: on_success   # regular jobs need one true if clause to run  
```

File: .gitlab/lassen-templates.yml
Comment (adrienbernede) [inline .gitlab/lassen-templates.yml:19]: See suggestion above

File: .gitlab/ruby-templates.yml
Comment (adrienbernede) [inline .gitlab/ruby-templates.yml:22]: See suggestion above

File: .gitlab/rznevada-templates.yml
Comment (adrienbernede) [inline .gitlab/rznevada-templates.yml:21]: See suggestion above

**Review by adrienbernede — CHANGES_REQUESTED** [review]
Rules can be confusing, I had to re-read the doc to remind myself of the exact way it works. Is my suggestion making sense to you ?

File: .gitlab/corona-templates.yml
Comment (kab163) [reply .gitlab/corona-templates.yml]: I think I have incorporated this in to my yaml files. Hopefully I will get these pipeline rules figured out soon! Thanks for the example!

Comment (adrienbernede) [reply .gitlab/corona-templates.yml]: There is a typo in my example I found in your code (my bad...): the proper syntax for "not equal" is `=~`.

Comment (kab163) [reply .gitlab/corona-templates.yml]: Oh I see haha I should have caught that!

File: .gitlab/corona-templates.yml
Comment (adrienbernede) [inline .gitlab/corona-templates.yml:20]: One again my bad, I appear to be a bit rusted and confused "loose string match" with "not equal":
```suggestion
    - if: '$LC_ZONE != "CZ"' #run only on CZ ...
```

File: .gitlab/lassen-templates.yml
Comment (adrienbernede) [inline .gitlab/lassen-templates.yml:18]:
```suggestion
    - if: '$LC_ZONE != "CZ"' #run only on CZ ...
```

File: .gitlab/ruby-templates.yml
Comment (adrienbernede) [inline .gitlab/ruby-templates.yml:20]:
```suggestion
    - if: '$LC_ZONE != "CZ"' #run only on CZ ...
```

**Review by adrienbernede — CHANGES_REQUESTED** [review]
Sorry for adding confusion...

File: .gitlab/rznevada-templates.yml
Comment (adrienbernede) [inline .gitlab/rznevada-templates.yml:20]:
```suggestion
    - if: '$LC_ZONE != "RZ"' #run only on RZ ...
```

File: .gitlab/ruby-templates.yml
Comment (adrienbernede) [inline .gitlab/ruby-templates.yml]:
```suggestion
    - if: '$LC_ZONE != "CZ"' #run only on CZ ...
```

**Review by adrienbernede — COMMENTED** [review]
Another fix

**Review by adrienbernede — APPROVED** [review]

**Comment by kab163** [comment]
> Another fix

Thanks! Looks great. Rules are confusing lol

**Comment by rhornung67** [comment]
@kab163 does this run automatically at some regular interval (e.g., nightly)? How does one trigger it manually?

**Comment by kab163** [comment]
> @kab163 does this run automatically at some regular interval (e.g., nightly)? How does one trigger it manually?

So the last couple things I still have to figure out for this PR includes that piece. I need to set up the mirror for the repo so that when changes are pushed to github, it triggers the RZ Gitlab pipeline. For now, it has to be done manually. If this branch were merged and you wanted to test `my_other_rajaperf_branch` on the RZ Gitlab, you have to do `git push rz-gitlab my_other_rajaperf_branch`. That will push any changes in that branch to the RZ Gitlab and sync stuff up.

**Comment by rhornung67** [comment]
The RZ Gitlab pipeline cannot be triggered from GitHub, as I understand it. So we will want to have Gitlab pull from GitHub and run this pipeline nightly. We also want to be able to run it manually from RZ Gitlab.

When you get everything working, please include instructions for running it in the PR summary. Thank you.

**Comment by kab163** [comment]
We are just waiting for a way to start up nightly runs. I have not heard back on the status of that... in the meantime, instructions for manual builds have been included in the PR summary.

**Comment by rhornung67** [comment]
Thanks @kab163 I'm still waiting on adrien to approve my radiuss-spack-configs PR so I can pull that into my PR branch and merge that.....

**Review by adrienbernede — APPROVED** [review]

**Comment by rhornung67** [comment]
@kab163 do you have time to switch this over to rzvernal to see if it works there? Note that you have to use 'gfx90a' as the target architecture.  Thank you.

**Comment by kab163** [comment]
> @kab163 do you have time to switch this over to rzvernal to see if it works there? Note that you have to use 'gfx90a' as the target architecture. Thank you.

Absolutely! I will add another pipeline.

**Comment by kab163** [comment]
Still need to iron out some kinks between spack and rzvernal, but at least triggering a pipeline on rzvernal works - https://rzlc.llnl.gov/gitlab/raja/rajaperf/-/jobs/285498

File: .gitlab/rzvernal-jobs.yml
Comment (kab163) [reply .gitlab/rzvernal-jobs.yml:10]: Yes. So in order to use a CCE compiler, we need v14.0.0 in order to use rocm@5.1.0... I have not tried with like rocm@4.5.2 and CCE@13.0.2 though... I found out that CCE@13.0.0 does not work though - with either rocm 5.1.0 or 4.5.2

Comment (kab163) [reply .gitlab/rzvernal-jobs.yml:10]: (And CCE@14.0.0 is still in the works... afaik)

Comment (rhornung67) [reply .gitlab/rzvernal-jobs.yml:10]: Right. I remember the story now. Thanks for keeping up with it.

Comment (kab163) [reply .gitlab/rzvernal-jobs.yml:10]: I just pulled in the radiuss updates for this thru RAJA so that I can use the updates in RAJAPerf. Once that's merged in, the rznevada tests will start passing again.. Still have to figure out some rzvernal + Github CI issues before that pipeline works, but the local build with the configuration I have will work.

**Comment by davidbeckingsale** [comment]
Hey @adrienbernede - yup that would be great, we would appreciate your help making those changes here!

**Comment by trws** [comment]
> @trws, not that I feel offended or something, but technically, Uberenv is not pinning the Spack version, the raja config file for Uberenv is ;).
👍 
My mistake, I missed that it was configured that way rather than pinned upstream.

**Pull request closed without merging**
